src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs

```diff
@@ -206,67 +201,44 @@ public void RemoveDelegate(Delegate d, bool wrapArgs = false)
                     return;
                 }
 
-                // Update wrapper or remove from collection
-                Delegate? newDelegate = Delegate.Remove(wrapper!.Delegate, d);
-                if (newDelegate != null)
-                {
-                    wrapper.Delegate = newDelegate;
-                }
-                else
-                {
-                    _delegateWrappers.RemoveAt(removeIdx);
-                }
-            }
+                newWrappers = new DelegateWrapper[wrappers.Length - 1];
+                wrappers.AsSpan(0, removeIdx).CopyTo(newWrappers);
+                wrappers.AsSpan(removeIdx + 1).CopyTo(newWrappers.AsSpan(removeIdx));
+            } while (!PublishNewWrappers(newWrappers, wrappers));
         }
 
         public void RemoveDelegates(Func<Delegate, bool> condition)
         {
-            lock (_delegateWrappers)
+            DelegateWrapper[] wrappers;
+            DelegateWrapper[] newWrappers;
+            do
             {
-                // Find delegate wrapper indexes. Iterate in reverse such that the list to remove is sorted by high to low index.
-                List<int> toRemove = new List<int>();
-                for (int i = _delegateWrappers.Count - 1; i >= 0; i--)
-                {
-                    DelegateWrapper wrapper = _delegateWrappers[i];
-                    Delegate[] invocationList = wrapper.Delegate.GetInvocationList();
-                    foreach (Delegate delegateMaybe in invocationList)
-                    {
-                        if (condition(delegateMaybe))
-                        {
-                            Delegate? newDelegate = Delegate.Remove(wrapper!.Delegate, delegateMaybe);
-                            if (newDelegate != null)
-                            {
-                                wrapper.Delegate = newDelegate;
-                            }
-                            else
-                            {
-                                toRemove.Add(i);
-                            }
-                        }
-                    }
-                }
-
-                foreach (int idx in toRemove)
-                {
-                    _delegateWrappers.RemoveAt(idx);
-                }
+                wrappers = _delegateWrappers;
+                List<DelegateWrapper> tmp = new(wrappers);
+                tmp.RemoveAll(w => condition(w.Delegate));
+                newWrappers = tmp.ToArray();
             }
+            while (!PublishNewWrappers(newWrappers, wrappers));
         }
 
         public object? Invoke(object[] args)
         {
             Debug.Assert(!Empty);
             object? result = null;
 
-            lock (_delegateWrappers)
+            DelegateWrapper[] wrappers = _delegateWrappers;
+            foreach (DelegateWrapper wrapper in wrappers)
```

Comment:
Nit: there's no need for the local here, it could just be:
```C#
foreach (DelegateWrapper wrapper in _delegateWrappers)
```
as was there previously, but if you prefer this style, it doesn't hurt